### ISSUE TITLE
Add option to allow use of undefined in function invocation context (no-undefined)

### ISSUE THREAD

**Issue by gajus** [eslint/eslint#8029]
[`no-undefined`](http://eslint.org/docs/rules/no-undefined) restricts use of `undefined`.

However, there are valid use cases for `undefined` –

1.) when the intention is to force function to use the default parameter value, e.g.

```js
> ((foo = 'bar', baz) => { console.log(foo, baz); })(undefined, 'foo')
< bar foo

```

2.) when using type annotations, e.g.

```js
const test = (root: void, parameters: ParameterType) => {};

// Here, `test` can be invoked only with first parameter being `undefined`.

```

In both cases, it is required to use `undefined` as a function invocation parameter.

Therefore, it would be nice to have an option `allowAsParameter` that'd allow this behaviour.

**Comment by platinumazure** [comment]
In both cases, does `void 0` not work?

In my view, the purpose of that rule is less to disallow undefined values and more to disallow use of an Identifier named "undefined". There are other ways to express the value.

**Comment by gajus** [comment]
```js
((foo = 'bar', baz) => { console.log(foo, baz); })(void 0, 'foo')
```

This looks a bit odd, but yeah, it works.

In my particular case, I have decided to disable `no-undefined` rule. Risk potential does not offset the odd syntax.

However, I understand what the rule attempts to achieve, and it would be illogical to allow use of undefined for parameter values.

**Comment by platinumazure** [comment]
Sounds good. Remember that you can always create a custom rule or plugin if
you really want to reimplement the rule with finer grained control over
parameters. Thanks for the discussion!
